sql/sql_path.cc

```diff
@@ -0,0 +1,412 @@
+    */
+    if (schema.length == cur_schema.length &&
+                        !memcmp(schema.str, cur_schema.str, cur_schema.length))
+      schema= thd->db;
+
+    if (!schema.str)
+      continue;
+
+    if (likely(!check_db_dir_existence(schema.str)))
+    {
+      const Lex_ident_db_normalized dbn=
+        thd->to_ident_db_normalized_with_error(schema);
+      if (dbn.str)
+      {
+        sp_name *spname= new (thd->mem_root) sp_name(dbn, name, false);
+        if (unlikely(!spname))
+          return true;
+
+        if (!sph->sp_find_routine_quick(thd, spname))
+        {
+          // found
+          if (dbn_out)
+            *dbn_out= dbn;
+          if (spname_out)
+            *spname_out= spname;
+
+          break;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+
+bool Sql_path::find_db_qualified(THD *thd,
+                                 sp_name *name,
+                                 const Sp_handler **sph,
+                                 Database_qualified_name *pkgname) const
+{
+  for (size_t i= 0; i < m_count; i++)
+  {
+    LEX_CSTRING schema= m_schemas[i];
+    /*
+      Resolve the schema name to the current database if required.
+    */
+    if (schema.length == cur_schema.length &&
+        !memcmp(schema.str, cur_schema.str, cur_schema.length))
+      schema= thd->db;
+
+    if (unlikely(!schema.str))
+      continue;
+    
+    if (likely(!check_db_dir_existence(schema.str)))
+    {
+      const Lex_ident_db_normalized dbn=
+        thd->to_ident_db_normalized_with_error(schema);
+      if (dbn.str)
+      {
+        if (!((*sph)->sp_find_qualified_routine(thd, dbn, name)))
+        {
+          pkgname->m_db= dbn;
+          pkgname->m_name= Lex_ident_routine(name->m_db);
+          *sph= (*sph)->package_routine_handler();
+          return name->make_package_routine_name(thd->mem_root, dbn,
+                                                 name->m_db, name->m_name);
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+
+Schema *Sql_path::find_first_internal_schema() const
+{
+  for (size_t i= 0; i < m_count; i++)
+  {
+    if (Schema *schema= Schema::find_by_name(m_schemas[i]))
+      return schema;
+  }
+
+  return nullptr;
+}
+
+
+void Sql_path::free()
+{
+  m_count= 0;
+
+  if (m_buffer)
+  {
+    my_free(m_buffer);
+    m_buffer= nullptr;
+    m_buffer_length= 0;
+  }
+}
+
+
+bool Sql_path::init()
+{
+  init_array();
+  free();
+
+  return false;
+}
+
+
+bool Sql_path::add_schema(const char *schema_str, size_t schema_len)
+{
+  DBUG_ASSERT(schema_str);
+
+  if (unlikely(m_count >= m_schemas.size()))
+    return true;
+  
+  /*
+    Disallow duplicate schema names. The intent here is to avoid
+    ambiguity in the order of schema resolution when the path
+    contains duplicate schema names.
+  */
+  for (size_t i= 0; i < m_count; i++)
+  {
+    if (unlikely(m_schemas[i].length == schema_len &&
+        !memcmp(m_schemas[i].str, schema_str, schema_len)))
```

Comment:
This is not correct, you need to compare normalized paths, not what user has specified.
```sql
   SET PATH='A,a'
```
would pass your check, but on Windows these schema names are equivalent.